src/java/org/apache/cassandra/repair/autorepair/RepairRangeSplitter.java

```diff
@@ -0,0 +1,396 @@
+import org.apache.cassandra.db.lifecycle.SSTableIntervalTree;
+import org.apache.cassandra.db.lifecycle.SSTableSet;
+import org.apache.cassandra.db.lifecycle.View;
+import org.apache.cassandra.dht.AbstractBounds;
+import org.apache.cassandra.dht.Range;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.io.sstable.format.SSTableReader;
+import org.apache.cassandra.io.sstable.format.big.BigTableScanner;
+import org.apache.cassandra.io.sstable.metadata.CompactionMetadata;
+import org.apache.cassandra.io.sstable.metadata.MetadataType;
+import org.apache.cassandra.service.StorageService;
+import org.apache.cassandra.utils.concurrent.Refs;
+
+import static org.apache.cassandra.repair.autorepair.AutoRepairUtils.split;
+
+/**
+ * In Apache Cassandra, tuning the repair ranges has three main goals:
+ * <p>
+ * 1. **Create smaller, consistent repair times**: Long repairs, such as those lasting 15 hours, can be problematic.
+ * If a node fails 14 hours into the repair, the entire process must be restarted. The goal is to reduce the impact
+ * of disturbances or failures. However, making the repairs too short can lead to overhead from repair orchestration
+ * becoming the main bottleneck.
+ * <p>
+ * 2. **Minimize the impact on hosts**: Repairs should not heavily affect the host systems. For incremental repairs,
+ * this might involve anti-compaction work. In full repairs, streaming large amounts of data—especially with wide
+ * partitions—can lead to issues with disk usage and higher compaction costs.
+ * <p>
+ * 3. **Reduce overstreaming**: The Merkle tree, which represents data within each partition and range, has a maximum size.
+ * If a repair covers too many partitions, the tree’s leaves represent larger data ranges. Even a small change in a leaf
+ * can trigger excessive data streaming, making the process inefficient.
+ * <p>
+ * Additionally, if there are many small tables, it's beneficial to batch these tables together under a single parent repair.
+ * This prevents the repair overhead from becoming a bottleneck, especially when dealing with hundreds of tables. Running
+ * individual repairs for each table can significantly impact performance and efficiency.
+ * <p>
+ * To manage these issues, the strategy involves estimating the size and number of partitions within a range and splitting
+ * it accordingly to bound the size of the range splits.
+ */
+
+public class RepairRangeSplitter implements IAutoRepairTokenRangeSplitter
+{
+    private static final Logger logger = LoggerFactory.getLogger(RepairRangeSplitter.class);
+
+    static final String SUBRANGE_SIZE = "bytes_per_assignment";
+    static final String PARTITION_COUNT = "partitions_per_assignment";
+    static final String TABLE_BATCH_LIMIT = "max_tables_per_assignment";
+    static final String MAX_BYTES_PER_SCHEDULE = "max_bytes_per_schedule";
+
+    private final int tablesPerAssignmentLimit;
+    private final long maxBytesPerScheduleBytes;
```

Comment:
could probably be just `maxBytesPerSchedule` (unless added Bytes to make unit more clear?)